### DISCUSSION TITLE
Upgrading to psycopg3 with AWS RDS and Bastion with SSL enabled

### DISCUSSION THREAD

**Discussion by nicholastulach** [psycopg/psycopg#270] [Q&A]
Hi,

I am trying to use psycopg3 with AWS RDS behind a Bastion tunnel. Our connections work with psycopg2 and i've got sample code working with psycopg3 to my local PostgreSQL server, but i cannot get a consistent connection to work with our Bastion/RDS setup with SSL. 

Sometimes this connection works (i'd say about 1/8th of the time), but most of the time it fails with the following error:

```
E           psycopg.OperationalError: connection failed: :1), port 5432 failed: could not receive data from server: Connection refused
E           could not send SSL negotiation packet: Connection refused
```

The Bastion host tunnels the port 5432 to the AWS RDS server and that works for psycopg2 (and occasionally works with psycopg3!)

Here's some sample code…

```
    with psycopg.connect(
        host=POSTGRES_HOST,
        user=POSTGRES_USER,
        port=POSTGRES_PORT,
        password=pg_pass,
        dbname=POSTGRES_DB,
        autocommit=True,
        client_encoding="UTF8",
        sslrootcert=cert,
        sslmode=POSTGRES_SSL_MODE,
    ) as conn:
        conn.autocommit = True
        logging.info("Connected to db; querying…")
        cur = conn.execute("SELECT * FROM mytable LIMIT 100;")
        logging.info(cur.fetchone())
```

Any help would be greatly appreciated!

**Comment by dvarrazzo** [comment]
Are you saying that the *exact* same connection string works reliably in psycopg 2 and not in 3?

That's kinda strange as they both use the same library to perform connection.

**Reply by nicholastulach** [reply]
Yes, the exact same connection information works with psycopg2 but not 3. I believe the clue is with the error it is throwing, the SSL negotiation is failing for some reason, but because it is non-deterministic, it is difficult to debug.

**Reply by nicholastulach** [reply]
Some more testing today. I ran the following code with `psycopg2` instead of 3 and it works _every time_. No weird SSL negotiation errors or anything. Just rock solid. :/

```
    with psycopg2.connect(
        host=POSTGRES_HOST,
        user=POSTGRES_USER,
        port=POSTGRES_PORT,
        password=pg_pass,
        dbname=POSTGRES_DB,
        #autocommit=True,
        client_encoding="UTF8",
        sslrootcert=cert,
        sslmode=POSTGRES_SSL_MODE,
    ) as conn:
        conn.autocommit = True
        logging.info("Connected to db; querying…")
        cur = conn.cursor()
        cur.execute("SELECT * FROM mytable LIMIT 100;")
        logging.info(cur.fetchone())
```

**Reply by dvarrazzo** [reply]
What versions of psycopg 2 and 3, and what binary/non binary option, are you using?

**Reply by nicholastulach** [reply]
psycopg2-binary 2.8.6
psycopg 3.0.11
 psycopg-c 3.0.11
 psycopg-pool 3.1.1

**Reply by nicholastulach** [reply]
(Testing on an Apple Silicon Mac, so i can't use `psycopg-binary` for v3.)

**Reply by dvarrazzo** [reply]
You can't use psycopg2-binary for v2 either. However psycopg2-binary has a fallback and installs from source if binary package fails to install.

So, it seems that you are using your system libpq both for psycopg2 and psycopg 3.

Psycopg 2 uses by default a blocking libpq function, but can be switched to using the same non-blocking one used by Psycopg 3 by registering a wait callback. So:

- please make sure you are using system libraries. `pip uninstall psycopg2-binary` and `pip install psycopg2`
- before connecting with psycopg2, call `psycopg2.extensions.set_wait_callback(psycopg2.extras.wait_select)`.
- does it fail? If so, there might be something broken in the way async connection is established.
- if it doesn't fail, it's stranger than already is.

**Reply by nicholastulach** [reply]
Okay, so i tried this with `psycopg` instead of `psycopg-c` and it works! So it seems like (maybe) it's a problem with `psycopg-c`?

**Reply by dvarrazzo** [reply]
Which makes me think: how is the tunnel established? There was some issue with some tunnel programs not working in async mode: see https://github.com/pahaz/sshtunnel/issues/135 (which might have been since solved, I don't know).

**Reply by nicholastulach** [reply]
We are using AWS ssm to a Bastion host, which i don't fully understand how it works because i am not a DevOps person. :D

**Comment by dvarrazzo** [comment]
> Okay, so i tried this with psycopg instead of psycopg-c and it works! So it seems like (maybe) it's a problem with psycopg-c?

That would be strange too. They use the same connection function from the libpq, except that the first call it [from python](https://github.com/psycopg/psycopg/blob/a928baabec2d6ddcc8459df1c8b1044a6fe90206/psycopg/psycopg/generators.py#L32-L61), the second [from Cython](https://github.com/psycopg/psycopg/blob/a928baabec2d6ddcc8459df1c8b1044a6fe90206/psycopg_c/psycopg_c/_psycopg/generators.pyx#L23-L58).

I wanted to ask you to add some debug prints. I see that there is a debugging leftover in cython which we can use. We can add similar info in Python. Can you add:
```diff
diff --git a/psycopg/psycopg/generators.py b/psycopg/psycopg/generators.py
index 35bbc432..e80cb259 100644
--- a/psycopg/psycopg/generators.py
+++ b/psycopg/psycopg/generators.py
@@ -35,6 +35,7 @@ def connect(conninfo: str) -> PQGenConn[PGconn]:
 
     """
     conn = pq.PGconn.connect_start(conninfo.encode())
+    logger.debug("connection started, status %s", conn.status)
     while 1:
         if conn.status == ConnStatus.BAD:
             encoding = conninfo_encoding(conninfo)
@@ -43,6 +44,7 @@ def connect(conninfo: str) -> PQGenConn[PGconn]:
             )
 
         status = conn.connect_poll()
+        logger.debug("connection polled, status %s", conn.status)
         if status == PollingStatus.OK:
             break
         elif status == PollingStatus.READING:
```

and then try to connect with the two alternative implementations. You can install `psycopg-c` and use the `PSYCOPG_IMPL` env var to choose the version to use.

```
$ PSYCOPG_IMPL=python python
>>> import logging
>>> logging.basicConfig(level=logging.DEBUG)
>>> logging.getLogger("psycopg").setLevel(logging.DEBUG)
>>> import psycopg
>>> print(psycopg.pq.__impl__)
python
>>> psycopg.connect()
DEBUG:psycopg.generators:connection started, status 3
DEBUG:psycopg.generators:connection polled, status 4
DEBUG:psycopg.generators:connection polled, status 0
<psycopg.Connection [IDLE] (database=piro) at 0x7f40455feaf0>
```

```
$ PSYCOPG_IMPL=c python
>>> import logging
>>> logging.basicConfig(level=logging.DEBUG)
>>> logging.getLogger("psycopg").setLevel(logging.DEBUG)
>>> import psycopg
>>> print(psycopg.pq.__impl__)
c
>>> psycopg.connect()
DEBUG:psycopg.adapt:connection started, status 3
DEBUG:psycopg.adapt:connection polled, status 4
DEBUG:psycopg.adapt:connection polled, status 0
<psycopg.Connection [IDLE] (database=piro) at 0x7f7838c60eb0>
```

What happens with your connection string?

**Reply by nicholastulach** [reply]
Question on setup for `psycopg-c`: Am i supposed to have both `psycopg` and `psycopg-c` installed? The instructions are unclear, but when i only install the latter, it fails, so i assume i need both packages (unlike for `psycopg-binary`).

**Reply by dvarrazzo** [reply]
You definitely need `psycopg`; on top of it you can add `psycopg-c`  (which is a speedup package) or `psycopg-binary` (which is the same of `psycopg-c`, but precompiled).

At import time, `psycopg.pq` will look for the `-c` package, then the `-binary` package, and use them if available, and will fall back on the Python implementation (which uses the libpq from Python via ctypes).

If you think that the [installation docs](https://www.psycopg.org/psycopg3/docs/basic/install.html) can be made clearer, suggestions are welcome :)

**Reply by nicholastulach** [reply]
> What happens with your connection string?

Can't seem to get this logging to work. Using poetry to manage my package installs.

But… annoyingly somehow the C version is now working for the same tests today. I did reinstall the packages via poetry, but it was not working before (for both me and another developer on a different computer). However it is still not working in our application code, so 🤷🏻‍♂️

**Reply by khushboovashi** [reply]
@dvarrazzo,

We are facing the same issue with pgAdmin 4. With psycopg2, it is working fine but not with psycopg3.

I have performed the tests that you asked for. Here is the result:

```
================================================================
>>> import logging
>>> logging.basicConfig(level=logging.DEBUG)
>>> logging.getLogger("psycopg").setLevel(logging.DEBUG)
>>> import psycopg
>>> print(psycopg.pq.__impl__)
c
>>> psycopg.connect("dbname=postgres")
<psycopg.Connection [IDLE] (user=khushboovashi database=postgres) at 0x10700fd50>
>>> psycopg.connect("dbname=postgres service=pg_13")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/khushboovashi/Projects/python_vm/psycopg/lib/python3.11/site-packages/psycopg/connection.py", line 728, in connect
    raise ex.with_traceback(None)
psycopg.OperationalError: connection failed: :1), port 61012 failed: could not receive data from server: Connection refused
could not send SSL negotiation packet: Connection refused

================================================================
>>> import logging
>>> logging.basicConfig(level=logging.DEBUG)
>>> logging.getLogger("psycopg").setLevel(logging.DEBUG)
>>> import psycopg
>>> print(psycopg.pq.__impl__)
python
>>> psycopg.connect("dbname=postgres service=pg_13")
DEBUG:psycopg.generators:connection started, status 2
DEBUG:psycopg.generators:connection polled, status 2
DEBUG:psycopg.generators:connection polled, status 3
DEBUG:psycopg.generators:connection polled, status 7
DEBUG:psycopg.generators:connection polled, status 3
DEBUG:psycopg.generators:connection polled, status 4
DEBUG:psycopg.generators:connection polled, status 0
<psycopg.Connection [IDLE] (host=localhost port=61012 user=khushboovashi database=postgres) at 0x1046c2990>
```

So, Python implementation is working fine but not C.

**Reply by dvarrazzo** [reply]
What is that strange `:1),` in the error message? How is the service `pg_13` defined?

**Reply by khushboovashi** [reply]
[pg_13]
host=localhost
port=61012
dbname=postgres

I have disabled IPV6 from pg_hba.conf purposefully.

**Reply by dvarrazzo** [reply]
This is very strange: because you are using the `c` extension, you should be binding to the same libpq and, as a consequence, to the same openssl library, of the `python` variant. By the look of it this isn't happening.

Could it be that you have several libpq installed in the system, and the one pointed at by pg_config is not the one that is then found by the dynamic linker?

What happens if you try with `psycopg[binary]`?

**Reply by khushboovashi** [reply]
Yes, I have multiple libpq installations on my system. Now, how would I ensure which libpq is being found by the dynamic linker?
With psycopg[binary], it is working fine.

**Reply by dvarrazzo** [reply]
On Linux, using the env var `LD_LIBRARY_PATH` to give the precedence to a path where you have libraries you want to find. That's usually useful for library developers, but, if you are starting needing this, it might be `$DEITY`'s way of telling you that your computer needs some cleanup...

Using the `ldd` program you can display what libpq and openssl library are linked at runtime by the `psycopg[c]` dynamic library. On my machine I see something like:

```
(.venv) piro@sherekhan:~/dev/psycopg3$ python -c "import psycopg; import psycopg_c; print(psycopg_c.pq)"
<module 'psycopg_c.pq' from '/home/piro/dev/psycopg3/.venv/lib/python3.10/site-packages/psycopg_c/pq.cpython-310-x86_64-linux-gnu.so'>

(.venv) piro@sherekhan:~/dev/psycopg3$ ldd /home/piro/dev/psycopg3/.venv/lib/python3.10/site-packages/psycopg_c/pq.cpython-310-x86_64-linux-gnu.so
	linux-vdso.so.1 (0x00007ffff63d3000)
	libpq.so.5 => /lib/x86_64-linux-gnu/libpq.so.5 (0x00007f642d082000)    <<<
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f642ce00000)
	libssl.so.3 => /lib/x86_64-linux-gnu/libssl.so.3 (0x00007f642cd5c000)  <<<
	libcrypto.so.3 => /lib/x86_64-linux-gnu/libcrypto.so.3 (0x00007f642c800000)
	libgssapi_krb5.so.2 => /lib/x86_64-linux-gnu/libgssapi_krb5.so.2 (0x00007f642d02e000)
	libldap-2.5.so.0 => /lib/x86_64-linux-gnu/libldap-2.5.so.0 (0x00007f642ccfd000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f642d13a000)
	libkrb5.so.3 => /lib/x86_64-linux-gnu/libkrb5.so.3 (0x00007f642c735000)
	libk5crypto.so.3 => /lib/x86_64-linux-gnu/libk5crypto.so.3 (0x00007f642ccce000)
	libcom_err.so.2 => /lib/x86_64-linux-gnu/libcom_err.so.2 (0x00007f642ccc8000)
	libkrb5support.so.0 => /lib/x86_64-linux-gnu/libkrb5support.so.0 (0x00007f642ccba000)
	liblber-2.5.so.0 => /lib/x86_64-linux-gnu/liblber-2.5.so.0 (0x00007f642cca9000)
	libsasl2.so.2 => /lib/x86_64-linux-gnu/libsasl2.so.2 (0x00007f642cc8e000)
	libgnutls.so.30 => /lib/x86_64-linux-gnu/libgnutls.so.30 (0x00007f642c54a000)
	libkeyutils.so.1 => /lib/x86_64-linux-gnu/libkeyutils.so.1 (0x00007f642cc87000)
	libresolv.so.2 => /lib/x86_64-linux-gnu/libresolv.so.2 (0x00007f642cc73000)
	libp11-kit.so.0 => /lib/x86_64-linux-gnu/libp11-kit.so.0 (0x00007f642c40f000)
	libidn2.so.0 => /lib/x86_64-linux-gnu/libidn2.so.0 (0x00007f642cc52000)
	libunistring.so.2 => /lib/x86_64-linux-gnu/libunistring.so.2 (0x00007f642c265000)
	libtasn1.so.6 => /lib/x86_64-linux-gnu/libtasn1.so.6 (0x00007f642c24d000)
	libnettle.so.8 => /lib/x86_64-linux-gnu/libnettle.so.8 (0x00007f642c207000)
	libhogweed.so.6 => /lib/x86_64-linux-gnu/libhogweed.so.6 (0x00007f642c1bf000)
	libgmp.so.10 => /lib/x86_64-linux-gnu/libgmp.so.10 (0x00007f642c13d000)
	libffi.so.8 => /lib/x86_64-linux-gnu/libffi.so.8 (0x00007f642cc43000)
```

The entries highlighted with `<<<` are libpq and libssl; this output shows that the ones linked are the system ones.

From your error message it seems that the libpq is failing to call the libssl and even failing to get its error message. Mine could be a long shot, but I think that are actually using a libpq that was compiled with a certain openssl library, but now it's linking with a different one.

The output of `ldd` is also affected by `LD_LIBRARY_PATH`, so you can actually using as a preview to know what will be linked at runtime.

**Comment by richstokes** [comment]
FWIW, I started (today) seeing this error on my local M1 MacBook, but it works fine in prod (intel). Not sure what changed, as it used to work fine locally too 🤔 

Lmk if there's any debug info I can collect to help with. tbh the multiple versions (c vs binary vs whatever) is confusing and not spent any time trying to figure that out so far.

**Reply by richstokes** [reply]
Seems really intermittent/non deterministic. Sometimes when I run it it's fine, other times it errors 🤔 

I'm using

```
psycopg                       3.1.3
psycopg-binary                3.1.3
psycopg-pool                  3.1.2
```

**Comment by JelteF** [comment]
I can reproduce this issue consistently enough that the SSL tests for pgbouncer never pass on Cirrus CI, which is using an ARM based MacOS runner. An example of a failed run can be found here: https://cirrus-ci.com/task/5261931737513984

It's failing with psycopg, psycopg-c and psycopg-binary in the same way

**Comment by vkocubinsky** [comment]
I catch the same issue after upgrade from python 3.10 to python 3.11. That is psycopg2 works but psycopg3 doesn't. I use PostgreSQL from docker  "image: bitnami/postgresql:14". 


```
In [1]: import psycopg2

In [2]: psycopg2.connect("postgresql://dev:1234@myhost:5432/mydb")
Out[2]: <connection object at 0x102467760; dsn: 'user=dev password=xxx dbname=mydb host=myhost port=5432', closed: 0>

In [3]: import psycopg

In [4]: psycopg.connect("postgresql://dev:1234@myhost:5432/mydb")
---------------------------------------------------------------------------
OperationalError                          Traceback (most recent call last)
Cell In[4], line 1
----> 1 psycopg.connect("postgresql://dev:1234@myhost:5432/mydb")

File ~/Library/Python/3.11/lib/python/site-packages/psycopg/connection.py:728, in Connection.connect(cls, conninfo, autocommit, prepare_threshold, row_factory, cursor_factory, context, **kwargs)
    723     rv = cls._wait_conn(
    724         cls._connect_gen(conninfo, autocommit=autocommit),
    725         timeout=params["connect_timeout"],
    726     )
    727 except e.Error as ex:
--> 728     raise ex.with_traceback(None)
    730 if row_factory:
    731     rv.row_factory = row_factory

OperationalError: connection failed: :18a0:6af2:b15d:b7c8), port 5432 failed: could not receive data from server: Connection refused

```

**Reply by dvarrazzo** [reply]
@vkocubinsky You must have a problem between ipv4 and ipv6 there. Probably the server is listening on ipv4 only but I can see that you are connecting on ipv6.

**Reply by vkocubinsky** [reply]
Thanks, I commented in issue https://github.com/psycopg/psycopg/issues/519. Not sure which place is better to discuss about the issue.
